In [1]:
import numpy as np

In [2]:
def gradient_descent(gradient, init_, learn_rate, n_iter = 50, tol = 1e-06):
    x = init_
    for _ in range(n_iter):
        delta = -learn_rate * gradient(x)
        if np.all(np.abs(delta) <= tol):
            break
        x += delta
    return round(x*1000)/1000

# Q1. (i)

$f'(x) = 2x + 3$

In [3]:
def gradient_1(x):
    return 2*x + 3

In [5]:
x = gradient_descent(gradient_1, 0, 0.2)
x

-1.5

In [16]:
f = (x)**2 + 3*x + 4
f

4.259056

The stationary point we got is (-1.5, 1.75)  
f''(x) at this point is 2 > 0, therefore this stationary point is the point of minima  

# Q1. (ii)

$f'(x) = 4x^3 - 6x + 2$

In [11]:
def gradient_2(x):
    return 4*x**3 - 6*x + 2

In [12]:
x = gradient_descent(gradient_2, 0, 0.2)
x

0.084

In [14]:
f = x**4 - 3*x**2 + 2*x
f

0.146881787136

the stationary point we got is (0.084, 0.14688)  
$f''(x) = 12x^2 - 6$  
the value of f''(x) at this point is -5.91 < 0, therefore this point is the point of maxima

we missed minima because of large step size  
decreasing learning rate to 0.02

In [17]:
x = gradient_descent(gradient_2, 0, 0.02)
x

-1.366

In [18]:
f = x**4 - 3*x**2 + 2*x
f

-4.848076206064

the stationary point we got is (-1.366, -4.848)  
$f''(x) = 12x^2 - 6$  
the value of f''(x) at this point is 16.39 > 0, therefore this point is the point of minima